faunadb/streams/client.py

```diff
@@ -0,0 +1,115 @@
+from faunadb.request_result import RequestResult
+from .events import parse_stream_request_result_or_none, Error
+from .errors import StreamError
+
+VALID_FIELDS = {"diff", "prev", "document", "action"}
+
+
+class Connection(object):
+    """
+    The internal stream client connection interface.
+    This class handles the network side of a stream
+    subscription.
+
+    Current limitations:
+    Python requests module uses HTTP1; hyper is used for HTTP/2
+    """
+    def __init__(self, client, expression, options):
+        self._client = client
+        self.options = options
+        self.conn = None
+        self._fields = None
+        if isinstance(self.options, dict):
+            self._fields = self.options.get("fields", None)
+        elif hasattr(self.options, "fields"):
+            self._fields = self.options.field
+        if isinstance(self._fields, list):
+            union = set(self._fields).union(VALID_FIELDS)
+            if union != VALID_FIELDS:
+                raise Exception("Valid fields options are %s, provided %s."%(VALID_FIELDS, self._fields))
+        self._state = "idle"
+        self._query = expression
+        self._data = to_json(expression).encode()
+        try:
+            self.conn = HTTP20Connection(
+                    self._client.domain, port=self._client.port, enable_push=True)
+        except Exception as e:
+            raise StreamError(e)
+
+    def close(self):
+        """
+        Closes the stream subscription by aborting its underlying http request.
+        """
+        if self.conn is None:
+            raise StreamError('Cannot close inactive stream subscription.')
+        self.conn.close()
+        self._state = 'closed'
+
+    def error(self):
+        """
+        Retrieves the error from the event loop in an async implementation.
+        """
+        return self.error
+
+    def subscribe(self, on_event):
+        """Initiates the stream subscription."""
+        if self._state != "idle":
+            raise StreamError('Stream subscription already started.')
+        try:
+            self._state = 'connecting'
+            headers = self._client.session.headers
+            headers["Authorization"] = self._client._auth_header()
+            if self._client._query_timeout_ms is not None:
+                    headers["X-Query-Timeout"] = str(self._client._query_timeout_ms)
+            headers["X-Last-Seen-Txn"] = str(self._client.get_last_txn_time())
+            start_time = time()
+            url_params = ''
+            if isinstance(self._fields, list):
+                url_params= "?%s"%(urlencode({'fields': ",".join(self._fields)}))
+            id = self.conn.request("POST", "/stream%s"%(url_params), body=self._data, headers=headers)
+            self._state = 'open'
+            self._event_loop(id, on_event, start_time)
+        except Exception as e:
+            if callable(on_event):
+                on_event(Error(e), None)
+            raise
```

Comment:
I think we should be purely callback based. Having multiple error handling mechanisms could be confusing.